Upgrade bot detection: Auto-block 500+ bots with smart detection

kuzmany opened the pull request:
## 🚀 Upgrade: Smarter Bot Detection for Better Analytics

This upgrade replaces Mautic's old hardcoded bot list with **automatic smart detection** that blocks 500+ bots without any setup needed.

### ✅ What You Get
- **Automatic bot blocking** - Works out of the box, no configuration needed
- **500+ bots detected** - Google, Facebook, Bing, and all major crawlers
- **Always up-to-date** - New bots are detected automatically
- **Better accuracy** - Smart detection vs simple text matching
- **Cleaner analytics** - No more bot traffic in your reports

### 🔧 What Changed
- **Removed old static bot list** - No more manual maintenance of 390+ hardcoded patterns
- **Added smart detection** - Uses Matomo's proven bot detection library
- **Fixed detection bug** - Bot detection now works properly (was broken before)
- **Improved UI** - Clear labels showing bot blocking is automatic

### 👥 For Users
- **Nothing to do** - Bot detection works automatically
- **Optional custom patterns** - Advanced users can still add custom bots if needed
- **Better user experience** - Clear UI explains what's happening

### 🛠️ Technical Details
- Uses Matomo DeviceDetector library for comprehensive bot detection
- Fixed missing `parse()` call that prevented bot detection from working
- Set `do_not_track_bots` to empty array (users can add custom patterns)
- Updated UI translations to guide users properly

**Result**: Better bot detection that works automatically and requires zero maintenance! 🎉

RCheesley commented:
@kuzmany conflicts to resolve here and we need to pull in the latest changes from the branch once that's done. Thanks!

kuzmany committed:
fix: make DeviceDetectorFactoryInterface optional in IpLookupHelper

- Add optional DeviceDetectorFactoryInterface to IpLookupHelper constructor
- Add bot detection using DeviceDetector in IP tracking
- Update service configuration to inject device detector factory
- Ensure backward compatibility with null checks

kuzmany committed:
fix: resolve PHPStan errors in IpLookupHelperTest

- Add proper type declarations for constructor parameters
- Pass all parameters to parent constructor to avoid unused parameter warnings

kuzmany committed:
fix: remove unnecessary constructor from IpLookupHelperTest

- Remove custom constructor that was causing PHPStan errors
- PHPUnit TestCase constructor only accepts one parameter
- setUp() method is sufficient for test initialization

kuzmany committed:
Merge branch '7.x' into add-device-detector

kuzmany committed:
Merge branch '7.x' into add-device-detector

andersonjeccel reviewed (changes requested):
I think it's not working...

<img width="1080" height="1919" alt="image" src="https://github.com/user-attachments/assets/fb8cf9c1-4e5e-4422-ac72-9c0cc0226d40" />

How to reproduce:

1. Open Contacts
2. Toggle anonymous visitors
3. Ensure the list is empty
4. Open Settings > System settings
5. Scroll down to find accordion for IP tracking and bots
6. Find Googlebot on the list and delete it
7. Save
8. Open Components > Landing pages
9. Create a new lp and save
10. Back to list view
11. Click on 3 dots, copy public page URL
12. Open incognito window
13. In chrome, open developer settings
14. Find the Network tab
15. Click to open advanced network settings
16. In User agent, uncheck "Use browser default"
17. Under the dropdown, find a select Googlebot Smartphone
18. Reload the anonymous tab
19. Paste the URL and hit Enter/return
20. Open Mautic's contact list
21. Reload to see

kuzmany committed:
Merge branch '7.x' into add-device-detector

kuzmany committed:
Remove hardcoded bot list and fix Matomo device detector bot detection

- Remove extensive hardcoded bot list from CoreBundle config.php
- Set do_not_track_bots to empty array to rely on Matomo device detector
- Fix missing parse() call in IpLookupHelper for proper bot detection
- Now uses Matomo's comprehensive and up-to-date bot detection instead

This allows users to add custom bot patterns if needed while leveraging
the superior bot detection capabilities of the Matomo device detector.

kuzmany committed:
Update bot detection UI labels for better user experience

- Make field label clear it's optional and for advanced users only
- Add reassuring tooltip explaining automatic bot detection works
- Emphasize 500+ bots are blocked automatically (Google, Facebook, etc.)
- Remove technical jargon and external links for cleaner UX
- Guide users that most don't need to add anything to this field

kuzmany commented:
@andersonjeccel good catch. I found the issue and update it. 
Also removed static bot list and this option is now only optional for advanced users
Bots detection is now automatically for all requests

<img width="1081" height="542" alt="image" src="https://github.com/user-attachments/assets/6dd82467-822c-4c8a-b0f8-1632c1720720" />

kuzmany committed:
Remove null default from DeviceDetectorFactory parameter

- Remove = null default value from DeviceDetectorFactory injection
- Makes dependency injection more explicit and clear
- Ensures factory is properly injected via service container
- Improves code clarity and prevents potential null issues

kuzmany committed:
Fix device detector bot check placement in loop

- Move device detector bot check outside the foreach loop
- Ensures bot detection runs once per request, not once per bot pattern
- Prevents unnecessary multiple executions of device detection
- Improves performance and correct logic flow

kuzmany committed:
Fix PHP 8.1 deprecation: reorder constructor parameters

- Move required DeviceDetectorFactory parameter before optional ipLookup
- Update service definition in config.php to match new parameter order
- Fixes "Required parameter follows optional parameter" deprecation
- Ensures proper dependency injection order in PHP 8.1+

kuzmany committed:
Fix test constructor parameter order for IpLookupHelper

- Update test to match new constructor parameter order
- Move DeviceDetectorFactory before optional ipLookup parameter
- Fixes type error in IpLookupHelperTest where null was passed for required parameter
- Ensures test compatibility with PHP 8.1 constructor parameter ordering

kuzmany commented:
Note:

The decreased read rate (estimated 20%) is expected and correct after implementing industry-standard bot filtering via this PR, which now accurately excludes 500+ bots

kuzmany committed:
Merge branch '7.x' into add-device-detector

Resolve conflict in IpLookupHelper.php by keeping both:
- DeviceDetectorFactoryInterface import from PR
- Request import and inline assignment pattern from 7.x

kuzmany committed:
Merge upstream/7.x into add-device-detector

Resolved conflict in CoreBundle/Config/config.php:
- Kept do_not_track_bots empty (PR uses device detector instead)
- Adopted upstream's rememberme_lifetime (90 days)
- Used upstream's formatting style

kuzmany committed:
Add link to Matomo bots list in tooltip

Added reference to the full bots list for users who want to see what bots are automatically detected.

andersonjeccel reviewed (approved):
It works now

<img width="1080" height="956" alt="image" src="https://github.com/user-attachments/assets/31189f79-7f4f-46bc-b25a-151e2a5c50a9" />

<img width="1920" height="929" alt="image" src="https://github.com/user-attachments/assets/780b9d37-0e0d-4219-8665-69bb6d5d1238" />

escopecz committed:
Merge branch '7.x' into add-device-detector

escopecz reviewed (approved):
The code looks good 👍  Thanks Zdeno!

escopecz commented:
The failing tests are resolved in https://github.com/mautic/mautic/pull/15872. Whoever can, please review it so I could keep merging PRs.

matbcvo committed:
Merge branch '7.x' into add-device-detector

RCheesley commented on app/bundles/CoreBundle/Tests/Unit/Helper/IpLookupHelperTest.php:
```diff
@@ -2,26 +2,57 @@
+
+    public function testDeviceDetectorBotsDetectionFalse(): void
+    {
+        $request = new Request([], [], [], [], [], ['REMOTE_ADDR' => '73.77.245.53']);
```
For reference, Sonarqube is flagging up the hard-coding of IP addresses here, I am marking it as safe, given it's used in a test.

kuzmany committed:
Merge branch '7.x' into add-device-detector

kuzmany committed:
Fix IpLookupHelper service wiring: add missing DeviceDetectorFactory argument

The constructor expects DeviceDetectorFactoryInterface as argument #4,
but the service config was passing mautic.ip_lookup (MaxmindDownloadLookup)
in that position. This caused all CI jobs to fail with a type error.

kuzmany committed:
Fix flaky IpLookupHelperTest by clearing static IP address cache

The static $ipAddresses cache in IpLookupHelper persists across tests.
testDeviceDetectorBotsDetectionTrue cached IP 73.77.245.52 as non-trackable,
which caused testIsRequestTrackableReturnsTrueForNormalRequest to fail
when it reused the same IP and got the stale cached result.

escopecz committed:
Merge branch '7.x' into add-device-detector
